cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/SparkProgramStatusMetricsProvider.java

```diff
@@ -61,53 +64,75 @@ public class SparkProgramStatusMetricsProvider implements MetricsProvider {
-            return new ExecutionMetrics[]{};
-          } else {
-            return metrics;
+    try {
+      return Retries.supplyWithRetries(
+        () -> {
+          ExecutionMetrics[] metrics;
+          HttpResponse applicationResponse;
+          try {
+            applicationResponse = doGet(applicationsURL);
+            String attemptId = extractAttemptId(applicationResponse.getResponseBodyAsString(), runIdStr);
+            if (Objects.nonNull(attemptId) && !attemptId.isEmpty()) {
+              HttpResponse stagesResponse;
+              String stagesURL = String.format("%s%s/%s/%s/stages", sparkHistoricBaseURL, sparkApplicationsEndpoint,
+                                               runIdStr, attemptId);
+              stagesResponse = doGet(stagesURL);
+              metrics = extractMetrics(stagesResponse.getResponseBodyAsString());
+              if (Objects.isNull(metrics)) {
+                emitMetrics(runId, startTime, false);
+                logger.error("Error during metrics extraction, received null response");
+                return new ExecutionMetrics[]{};
+              } else {
+                emitMetrics(runId, startTime, true);
+                return metrics;
+              }
+            } else {
+              throw new RetryableException("Error during attemptId extraction");
+            }
+          } catch (IOException e) {
+            logger.warn("Error retrieving application response", e);
+            throw new RetryableException(e);
           }
-        } else {
-          throw new RetryableException("Error during attemptId extraction");
-        }
-      } catch (IOException e) {
-        logger.warn("Error retrieving application response", e);
-        throw new RetryableException(e);
-      }
-    }, RetryStrategies.fromConfiguration(this.cConf, Constants.Spark.SPARK_METRICS_PROVIDER_RETRY_STRATEGY_PREFIX),
-                                     t -> t instanceof RetryableException);
+        }, RetryStrategies.fromConfiguration(this.cConf, Constants.Spark.SPARK_METRICS_PROVIDER_RETRY_STRATEGY_PREFIX),
+        t -> t instanceof RetryableException);
+    } catch (Exception e) {
+      emitMetrics(runId, startTime, false);
+      logger.error("Retries failed for extracting metrics. ", e);
+      return new ExecutionMetrics[]{};
+    }
+  }
+
+  private void emitMetrics(ProgramRunId runId, long startTime, boolean success) {
+    Map<String, String> metricTags = new HashMap<>();
+    metricTags.put(Constants.Metrics.Tag.NAMESPACE, runId.getNamespace());
+    metricTags.put(Constants.Metrics.Tag.PROGRAM, runId.getProgram());
+    metricTags.put(Constants.Metrics.Tag.APP, runId.getApplication());
+    metricTags.put(Constants.Metrics.Tag.STATUS, success ? "success" : "failure");
+    metricsCollectionService.getContext(metricTags)
+      .gauge(Constants.Metrics.ProgramEvent.METRICS_FETCH_LATENCY_MS, System.currentTimeMillis() - startTime);
```

Comment:
nit: Can we change the name of this to `SPARK_METRICS_FETCH_LATENCY_MS` for clarity?

Reply:
Done